### PR TITLE
adding HMAC-secret support

### LINKED ISSUE #74: Support Windows Hello
## Expected Behavior
OpenSK can be used for login on Windows Hello.

## Actual Behavior
The required FIDO2 extension HMAC-secret is not implemented yet.

## Steps to Reproduce the Problem

1. Set a Windows 10 machine up.
1. Try passwordless with OpenSK

## Specifications

- Version: FIDO specification 2.0+
- Platform: Windows 10

### PR DESCRIPTION
**Pull request by kaczmarczyck**
Fixes #74 

This is the extension implementation for HMAC-secret. This is required for Windows Hello passwordless logins.

This is still WIP, the extension has not been tested on Windows 10 yet.

### CODE REVIEW

**Comment by kaczmarczyck** [comment]
I forgot to set the extension flag properly. Fixed and tested on Windows 10.

File: src/ctap/mod.rs [hunk 1]
```diff
@@ -105,6 +106,63 @@ fn check_pin_auth(hmac_key: &[u8], hmac_contents: &[u8], pin_auth: &[u8]) -> boo
     )
 }
 
+// Decrypts the HMAC secret salt(s) that were encrypted with the shared secret.
+// The credRandom is used as a secret to HMAC those salts.
+// The last step is to re-encrypt the outputs.
+pub fn encrypt_hmac_secret_output(
+    shared_secret: &[u8; 32],
+    salt_enc: Vec<u8>,
```
Comment (gendx) [inline src/ctap/mod.rs]: Is there any reason to prefer a Vec here as opposed to a slice? I don't see any Vec-specific functions being used.

File: src/ctap/data_formats.rs [hunk 2]
```diff
@@ -220,6 +220,78 @@ impl TryFrom<&cbor::Value> for Extensions {
     }
 }
 
+impl From<Extensions> for cbor::Value {
+    fn from(extensions: Extensions) -> Self {
+        cbor_map_btree!(extensions
+            .0
+            .iter()
+            .map(|(key, value)| (cbor_text!(key), value.clone()))
```
[the remark this answers is not in the document]
Comment (kaczmarczyck) [reply src/ctap/data_formats.rs]: Done

File: src/ctap/data_formats.rs [hunk 3]
```diff
@@ -1005,6 +1089,16 @@ mod test {
             ..credential
         };
 
+        assert_eq!(
+            PublicKeyCredentialSource::try_from(cbor::Value::from(credential.clone())),
+            Ok(credential.clone())
+        );
+
+        let credential = PublicKeyCredentialSource {
+            cred_random: Some([0x00; 32].to_vec()),
```
[the remark this answers is not in the document]
Comment (kaczmarczyck) [reply src/ctap/data_formats.rs]: Done

File: src/ctap/mod.rs [hunk 4]
```diff
@@ -362,6 +421,22 @@ where
             return Err(Ctap2StatusCode::CTAP2_ERR_UNSUPPORTED_ALGORITHM);
         }
 
+        let use_hmac_extension = if let Some(extensions) = extensions {
+            extensions.has_make_credential_hmac_secret()?
+        } else {
+            false
+        };
```
[the remark this answers is not in the document]
Comment (kaczmarczyck) [reply src/ctap/mod.rs]: Done

File: src/ctap/mod.rs [hunk 5]
```diff
@@ -84,6 +84,7 @@ pub const ENCRYPTED_CREDENTIAL_ID_SIZE: usize = 112;
 const UP_FLAG: u8 = 0x01;
 const UV_FLAG: u8 = 0x04;
 const AT_FLAG: u8 = 0x40;
+const ED_FLAG: u8 = 0x80;
```
[the remark this answers is not in the document]
Comment (kaczmarczyck) [reply src/ctap/mod.rs]: Done

File: src/ctap/mod.rs [hunk 6]
```diff
@@ -527,6 +611,16 @@ where
             }
         }
 
+        let get_assertion_hmac_secret_input = if let Some(extensions) = extensions {
+            extensions.get_assertion_hmac_secret().transpose()?
+        } else {
+            None
+        };
```
[the remark this answers is not in the document]
Comment (kaczmarczyck) [reply src/ctap/mod.rs]: Done

File: src/ctap/mod.rs [hunk 7]
```diff
@@ -590,7 +687,37 @@ where
...
+                // Again, hard to tell what the correct error code here is.
+                return Err(Ctap2StatusCode::CTAP2_ERR_UNSUPPORTED_EXTENSION);
+            }
+
+            let encrypted_output = if let Some(cred_random) = &credential.cred_random {
+                encrypt_hmac_secret_output(&shared_secret, salt_enc, cred_random)?
+            } else {
+                // This happens because the credential was not created with HMAC-secret.
+                return Err(Ctap2StatusCode::CTAP2_ERR_UNSUPPORTED_EXTENSION);
+            };
```
[the remark this answers is not in the document]
Comment (kaczmarczyck) [reply src/ctap/mod.rs]: Done

File: src/ctap/mod.rs (hunk 1 above)
Comment (kaczmarczyck) [reply src/ctap/mod.rs]: The reason was that the input is a Vec, so I just wanted to move it into the function. A slice should be good, too.

File: src/ctap/mod.rs [hunk 8]
```diff
@@ -1294,4 +1520,32 @@ mod test {
                 .is_none());
         }
     }
+
+    #[test]
+    fn test_encrypt_hmac_secret_output() {
+        let shared_secret = [0x55; 32];
+        let salt_enc = vec![0x5E; 32];
+        let cred_random = vec![0xC9; 32];
```
[the remark this answers is not in the document]
Comment (kaczmarczyck) [reply src/ctap/mod.rs]: Changed together with the comment below, the function now takes slices.

**Pull request merged**